File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque
dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]
def bfs(x, y, n, m, maps) :
    queue = deque()
    queue.append((x, y))
    
    while queue :
        x, y = queue.popleft()
        for i in range(4) :
            nx = x + dx[i]
            ny = y + dy[i]
            
            if 0 <= nx < n and 0 <= ny < m :
                if maps[nx][ny] == 1 :
                    maps[nx][ny] = maps[x][y] + 1
                    queue.append((nx, ny))

def solution(maps):
    answer = 0
    n = len(maps)
    m = len(maps[0])
    
    bfs(0, 0, n, m, maps)
    
    answer = maps[n - 1][m - 1]
    if answer == 1 :
        return -1
    
    
    return answer

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    public static int[] dx = new int[]{-1, 1, 0, 0};
    public static int[] dy = new int[]{0, 0, -1, 1};

    public static int solution(int[][] maps) {
        int n = maps.length;
        int m = maps[0].length;

        ArrayDeque<Integer[]> queue = new ArrayDeque<>();
        queue.addLast(new Integer[]{0, 0});

        while (!queue.isEmpty()) {
            Integer[] poll = queue.pollFirst();
            Integer x = poll[0];
            Integer y = poll[1];

            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];

                if (0 <= nx && nx < n && 0 <= ny && ny < m) {
                    if (maps[nx][ny] == 1) {
                        maps[nx][ny] = maps[x][y] + 1;
                        queue.addLast(new Integer[]{nx, ny});
                    }
                }
            }
        }

        return maps[n - 1][m - 1] != 1 ? maps[n - 1][m - 1] : -1;
    }
}